File: src/main/java/br/com/rotaractsorocabaleste/rotasales/sale/SaleController.java
package br.com.rotaractsorocabaleste.rotasales.sale;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.io.ByteArrayOutputStream;
import java.util.UUID;

@RestController
@RequestMapping("/sales")
@RequiredArgsConstructor
public class SaleController {

    private final SaleService saleService;

    @GetMapping
    public Page<GetSalesVO> findMySales(@RequestParam(required = false) final UUID eventId,
                                        @RequestParam(required = false) final String buyer,
                                        @RequestParam(required = false) final Boolean donation,
                                        @RequestParam(required = false, defaultValue = "0") final int page,
                                        @RequestParam(required = false, defaultValue = "10") final int limit) {
        return saleService.findMySales(eventId, buyer, donation, page, limit);
    }

    @PostMapping
    public SaleDTO create(@RequestBody @Valid final SaleDTO saleDTO) {
        return saleService.create(saleDTO);
    }

    @GetMapping(value = "/receipt/{saleId}", produces = MediaType.APPLICATION_PDF_VALUE)
    public byte[] getSaleReceiptInPdf(@PathVariable final UUID saleId) {
        return saleService.getSaleReceiptInPdfStream(saleId);
    }

    @GetMapping("/{saleId}")
    public SaleDTO findSaleById(@PathVariable final UUID saleId) {
        return saleService.findDTOById(saleId);
    }

    @PutMapping
    public void updateSale(@RequestBody @Valid final SaleDTO saleDTO) {
        saleService.updateSale(saleDTO);
    }

    @PatchMapping
    public void changeSaleStatus(@RequestBody @Valid final ChangeSaleStatusDTO changeSaleStatusDTO) {
        saleService.changeSaleStatus(changeSaleStatusDTO);
    }

    @DeleteMapping("/{saleId}")
    public void deleteById(@PathVariable final UUID saleId) {
        saleService.deleteById(saleId);
    }

}

File: src/main/kotlin/br/com/rotaractsorocabaleste/rotasales/entrypoint/controller/SaleController.kt
package br.com.rotaractsorocabaleste.rotasales.entrypoint.controller

import br.com.rotaractsorocabaleste.rotasales.core.entity.Sale
import br.com.rotaractsorocabaleste.rotasales.core.service.SaleService
import br.com.rotaractsorocabaleste.rotasales.core.vo.PatchSaleStatusRequestVO
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.security.Principal
import java.util.*

@RestController
@RequestMapping("/sales")
class SaleController(
    private val saleService: SaleService,
    private val logger: Logger = LoggerFactory.getLogger(SaleController::class.java)
) {

    @GetMapping
    fun getLoggedInUserSales(principal: Principal): ResponseEntity<List<Sale>> {
        val ret = saleService.getLoggedInUserSales()

        logger.info("Getting all sales for seller=${principal.name}")

        return ResponseEntity.ok(ret)
    }

    @PostMapping
    fun create(@RequestBody sale: Sale): ResponseEntity<Any> {
        logger.info("Received request for create a sale, sale for=${sale.customer}")

        saleService.create(sale)

        logger.info("New sale saved, sale for=${sale.customer}")

        return ResponseEntity.created(URI.create("/sales")).build()
    }

    @PutMapping
    fun update(@RequestBody sale: Sale): ResponseEntity<Any> {
        logger.info("Received request for update a sale, sale for=${sale.customer}")

        saleService.update(sale)

        logger.info("Sale updated, sale for=${sale.customer}")

        return ResponseEntity.ok().build()
    }

    @DeleteMapping("/{saleId}")
    fun delete(@PathVariable saleId: UUID): ResponseEntity<Any> {
        saleService.delete(saleId)

        logger.info("Sale deleted, saleId=${saleId}")

        return ResponseEntity.ok().build()
    }

    @PatchMapping
    fun patchSaleStatus(@RequestBody patchSaleStatusRequestVO: PatchSaleStatusRequestVO): ResponseEntity<Any> {
        logger.info("Received request for change saleId=${patchSaleStatusRequestVO.saleId}} to status=${patchSaleStatusRequestVO.status}")

        saleService.patchSaleStatus(patchSaleStatusRequestVO)

        return ResponseEntity.ok().build()
    }

}